fix: Create new StudioModal

TomasEng opened the pull request:
## Description

### Summary of changes
- Replaced the old `StudioModal` component with a new one built on top of the one from The Design System
- Updated usages of the component within `dashboard` and `app-development`
- Moved the old component to `resourceadm` and named it `ReasourceadmModal`
- Cleaned up some tests

### Description of the new component – differences from The Design System
Just like [the component in The Design System](https://storybook.designsystemet.no/?path=/docs/komponenter-modal--docs), this component exposes `Root`, `Trigger` and `Dialog` components, of which only the `Dialog` component is required.

It does not expose the `Header`, `Content` and `Footer` components. The corresponding elements are used similarly across Studio and therefore built-in in the `StudioModal.Dialog` component to make it easier to use.

The `StudioModal.Trigger` component uses `StudioButton` and accepts the same props.

### Demo of the new component
Code:
```
<StudioModal.Root>
  <StudioModal.Trigger>Open</StudioModal.Trigger>
  <StudioModal.Dialog
    closeButtonTitle='Close'
    footer='Test'
    heading='Lorem ipsum dolor sit amet'
    icon={<HouseIcon/>}
  >
    Donec suscipit et nulla vel iaculis. Aliquam in tellus quis sem malesuada rhoncus non vitae lorem. Morbi sed dolor
    pulvinar velit eleifend eleifend. Ut molestie ex neque, vel elementum odio fringilla vel. Nunc consectetur urna eget
    ex congue, in accumsan dui vestibulum. Donec aliquet massa vitae nibh mollis, nec viverra eros venenatis. Sed ac
    accumsan enim. Quisque elementum elit sollicitudin interdum luctus. Fusce lacinia mattis consectetur. Aenean
    vehicula orci vitae dolor imperdiet scelerisque. Curabitur hendrerit gravida erat ut tincidunt. Proin dui ante,
    euismod vehicula finibus vitae, eleifend ac dolor. Etiam lobortis neque eget elit varius, quis laoreet est laoreet.
    Nunc ac congue turpis, non lobortis nisi. Fusce lectus leo, porta vitae ligula ut, sagittis pretium eros.
  </StudioModal.Dialog>
</StudioModal.Root>
```

Result (when opened):
![image](https://github.com/user-attachments/assets/adeee739-377a-477c-bd96-cb5f0641b62b)

### Differences from the old component
In contrast to the old component, the new one renders a native `<dialog>` HTML element, resulting in a more semantically correct DOM. This element has natively built-in open/closed state handling, so we don't need to handle that in React. Instead, we can use the  element's `showModal` and `close` methods. These can be accessed indirectly by the `StudioModal.Trigger` element or directly with a `ref`.

One drawback of using `<dialog>` is that it is not (yet) fully supported by JSDOM, making it difficult to test some of its features with React Testing Library. On the other hand, it's better to use the solutions provided by The Design System than to build our own.

## Related Issue
- #11045

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)

> [!NOTE]
> There are some new lines not covered by tests according to Codecov. One of them is in the new component; this line is not possible to test with React Testing Library due to JSDOM limitations. There rest are already existing files. I have made an effort to improve some of the tests, but I haven't added anything new as I consider it's out of scope of this issue.

framitdavid commented on frontend/packages/shared/src/components/GiteaHeader/VersionControlButtons/components/ShareChangesPopover/CommitAndPushContent/FileChangesInfoModal/FileChangesInfoModal.module.css:
```diff
@@ -1,33 +1,11 @@
-.diffStatus {
-  padding: var(--fds-spacing-3);
+.dialog .table .fileStatusCell {
+  width: 1%;
```
Would this scale better, or am I misunderstanding why it's set to 1%?

```suggestion
  width: min-content;
```

TomasEng replied:
I know this is ugly, but it was the best trick I could come up with. `min-content` doesn't work on table cells.

framitdavid replied:
Alright, resolved! ☺️

framitdavid reviewed (commented):
Great work @TomasEng! Few comments only. 😊

TomasEng commented on frontend/app-development/types/SettingsModalHandle.ts:
```diff
@@ -0,0 +1,5 @@
+import type { SettingsModalTab } from './SettingsModalTab';
+
+export type SettingsModalHandle = {
```
It's also used in the `SettingsModalContext`, which spans the whole app.

TomasEng commented on frontend/libs/studio-components/package.json:
```diff
@@ -19,15 +19,14 @@
-    "@storybook/react": "^8.0.4",
-    "@storybook/react-webpack5": "^8.0.4",
-    "@storybook/test": "^8.0.4",
+    "@storybook/blocks": "8.2.7",
```
I think I just couldn't make things work as in the Storybook documentation, so I updated it to the latest version.
I may be wrong, but it also seems like this is not picked up by Renovate.

framitdavid replied:
Thanks! Keep it as is in this PR. I will follow to renovate issue you mentioned by describing the problem within an already existing issue.

TomasEng commented on frontend/app-development/layout/SettingsModalButton/SettingsModal/SettingsModal.tsx:
```diff
@@ -19,17 +18,20 @@ import { PolicyTab } from './components/Tabs/PolicyTab';
-};
-
-export const SettingsModal = ({ isOpen, onClose, defaultTab }: SettingsModalProps): ReactNode => {
+export const SettingsModal = forwardRef<SettingsModalHandle, {}>(({}, ref): ReactNode => {
```
Fikset!

framitdavid reviewed (approved):
Rock solid! 🙌

Annikenkbrathen commented:
I tested it in the dev environment, and on my end, there was something strange with the frame when I went to "local changes." It must be a bug (
<img width="729" alt="Skjermbilde 2024-08-19 kl  11 59 04" src="https://github.com/user-attachments/assets/5ef3bc22-ce72-47c3-948d-3895c960eba7">
see the image).
<img width="748" alt="Skjermbilde 2024-08-19 kl  11 58 29" src="https://github.com/user-attachments/assets/bc2df409-25e8-4dc1-86fc-5368c30f8ccc">

I also tested deleting my last changes, and the same problem with the small space occurred. But I also thought it was a bit strange that the modal doesn't close after you delete. Is that intentional? If you're not supposed to close both modals, at least the modal that appears when you click "delete my changes" should close.

TomasEng commented:
Thanks for testing, @Annikenkbrathen!

> I tested it in the dev environment, and on my end, there was something strange with the frame when I went to "local changes."

It looked as expected when I tested it with Chrome, also when making the window small:
![image](https://github.com/user-attachments/assets/56a77198-d5b0-4828-8e5d-28c2d1b1e249)
Which browser are you using? Does this happen with all the modals?

> I also tested deleting my last changes, and the same problem with the small space occurred. But I also thought it was a bit strange that the modal doesn't close after you delete. Is that intentional? If you're not supposed to close both modals, at least the modal that appears when you click "delete my changes" should close.

You're right, the modal should close after I delete. I'll look into it.

Annikenkbrathen commented:
> Thanks for testing, @Annikenkbrathen!
> 
> > I tested it in the dev environment, and on my end, there was something strange with the frame when I went to "local changes."
> 
> It looked as expected when I tested it with Chrome, also when making the window small: ![image](https://private-user-images.githubusercontent.com/29770305/359085762-56a77198-d5b0-4828-8e5d-28c2d1b1e249.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwNjc5OTMsIm5iZiI6MTcyNDA2NzY5MywicGF0aCI6Ii8yOTc3MDMwNS8zNTkwODU3NjItNTZhNzcxOTgtZDViMC00ODI4LThlNWQtMjhjMmQxYjFlMjQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE5VDExNDEzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIzNjQzNWRiMGY1NTJjNGU4Mjg2ZjFiNjc2NGIxM2MxY2Y1NmI4Nzc1ZGM0ZTNhZGE0NDM1MjIwMzZkYjg1YTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.18Jcoq3wSYIJPMy8gku6KNPUci3bfkDdU89C8KFh-2g) Which browser are you using? Does this happen with all the modals?

I use safari 😋
Now everyone looked as espected again, weird!
I checked settings modal, local changes and this! But it looks good now
<img width="1035" alt="Skjermbilde 2024-08-19 kl  13 44 14" src="https://github.com/user-attachments/assets/8429b535-e87c-4165-8f01-9485db20870a">

> 
> > I also tested deleting my last changes, and the same problem with the small space occurred. But I also thought it was a bit strange that the modal doesn't close after you delete. Is that intentional? If you're not supposed to close both modals, at least the modal that appears when you click "delete my changes" should close.
> 
> You're right, the modal should close after I delete. I'll look into it.
 👍🏼

TomasEng commented:
> I use safari Now everyone looked as espected again, weird! I checked settings modal, local changes and this! But it looks good now

This image looks like the old version (you can tell by the inconsistent spacings). It seems like there's another branch in the `dev` environment now. I'll deploy this one again when I have fixed the deletion issue. In the meantime, could you please verify that this is not happening in [the design system component](https://storybook.designsystemet.no/?path=/docs/komponenter-modal--docs)? @Annikenkbrathen

mlqn commented on frontend/libs/studio-components/src/components/StudioModal/StudioModalDialog/StudioModalDialog.module.css:
```diff
@@ -0,0 +1,39 @@
+
+.content {
+  padding: 0;
+  flex: 1;
```
I'm not sure if it's the best solution, but deleting this line seems to fix the height issue
```suggestion
```

TomasEng replied:
Nice catch! The reason for the `flex: 1` there is to make the `.content` element take up the full height. Without it, the settings modal looks like this:
![image](https://github.com/user-attachments/assets/48ee98f9-0a87-4ad4-ba45-4d5176e1242d)
Notice that the grey area does not reach down to the bottom.

However, since we know that this causes problems for Safari, I'll see if I can find another solution without using `flex`.

mlqn replied:
@TomasEng I've tried different solutions, but the one that worked best on both Safari and Chrome was to use `flex: 1` on the `.content` class, as you did, but only for the settings modal, as it appears to work well on Safari only if the modal has a height set (e.g., height: 80vh).
A solution that might not be the most optimal but that works could be to apply `flex: 1` to the `.content` class only for the settings modal by passing something like a `contentClassName`:

https://github.com/Altinn/altinn-studio/compare/ny-modal...ny-modal-test

mlqn replied:
Another solution could be to pass the header and the content as children to the modal, but this would require more work 🤷‍♂️ :

```
<StudioModal>
        <StudioModal.Header icon={<TrashIcon />} closeButtonTitle={}>...</StudioModal.Header>
        <StudioModal.Content className={classes.modalContent}>...</StudioModal.Content>
</StudioModal>
```

TomasEng replied:
Although the second suggestion is more aligned with how it is done in The Design System, I think it's too verbose for our use cases. It will just make the code harder to read. Therefore I think the `contentClassName` prop is a much better option.

How will the scrolling appear in the modals without `flex: 1`? If I remember correctly, I had some issues with the scrolling when I tried to remove it (which is why I reverted my first fix).

mlqn replied:
Agreed. The scrolling seems to work as exepected. Here is how it looks on both Safari and Chrome:

## Safari

### Modals without `flex: 1`

<table>

<tr>
<td>

https://github.com/user-attachments/assets/d13176d3-4526-4c55-8c41-14a37d4f6e3d

</td>
<td>

https://github.com/user-attachments/assets/5f2d32c5-2ef6-407e-8da9-20e4b6b99a55

</td>
<td>

https://github.com/user-attachments/assets/cff5a376-6ede-49ed-8766-78da64ae8ba9

</td>
</tr>
</table>

### Modal with `flex: 1`

The sidebar on the left gets an extra horizontal scrollbar, but this issue doesn't come from this PR, as the scrollbar is also visible with the old modal in production. I tried to fix it, but I couldn't find a solution that works well on both Chrome and Safari 🤷‍♂️ 

https://github.com/user-attachments/assets/bab19bea-7c66-4d94-8831-d9b43d27d9e7

## Chrome

### Modals without `flex: 1`

<table>

<tr>
<td>

[chrome-clone.webm](https://github.com/user-attachments/assets/d65bf49f-a4bc-40e4-ada6-d3dfd64bc3d3)

</td>
<td>

[chrome-diff.webm](https://github.com/user-attachments/assets/4eb498cc-4831-4858-a167-41db161efbe7)

</td>
<td>

[chrome-changes.webm](https://github.com/user-attachments/assets/1e6e9583-9514-42fc-a44e-1226f873273b)

</td>
</tr>
</table>

### Modal with `flex: 1`

[chrome-settings.webm](https://github.com/user-attachments/assets/94f81788-23d5-40a5-a6f4-3786381d2e97)

TomasEng replied:
Great work! I'll merge it in.

mlqn reviewed (commented):
Tested with Safari in DEV.

- The MakeCopy modal and the LocalChanges modals have a height issue. This seems to be the same issue we encountered with the previous modal (see https://github.com/Altinn/altinn-studio/pull/12116).

https://github.com/user-attachments/assets/aa672b43-77ce-4549-81a5-c58bcd957d93

https://github.com/user-attachments/assets/99cffc09-39d5-4090-ae1a-0b2b03a20187

- The FileChangesInfo modal works well, but there’s a small issue: you can see the files behind the header when scrolling

https://github.com/user-attachments/assets/1469a5ef-c699-4310-ac27-4c2cd7d37c0e

- The settings modal works well 👍 

https://github.com/user-attachments/assets/41ed7a6b-91f4-4d96-b53a-d107c17637a6

TomasEng commented:
Thanks for debugging, @mlqn! I'll try to find an alternative solution without using `flex`. When it comes to the file changes table, I suspect it may be a design system issue, but I'll look into that as well. Actually I don't think we need that table header to be sticky at all.

TomasEng commented:
I have now fixed the Safari problem by removing `flex: 1`, as suggested by @mlqn. In order to make the left panel appear correctly, I have applied the background to the modal element instead using `linear-gradient`.

Other fixes:
- Set correct levels to the headings in the settings dialog (since the design system modal uses `h2` as the top level heading)
- Removed sticky header effect from the file changes table
- Made the settings dialog smaller on large screens

TomasEng commented:
I have now updated this PR with a fix to the issue that made it impossible to open the settings modal after the user has opened it from the process page.
- Added a Playwright test that reproduced the issue. 
- Added some missing `useCallback`s.
- Switched the `StudioModal.Trigger` in `SettingsModalButton` with a regular `StudioButton`, since the issue was partly caused by the trigger not being connected to the dialog anymore when the dialog was rerendered.

TomasEng commented:
Pushed another fix now, related to reopening the local changes modal. It seems like there is some problem with handling `ref`s in the `Modal.Trigger` component. I replaced it with a regular `StudioButton`, which is fine here since we already have a defined `ref` to the `dialog` element in the parent component (probably this is also what causes problems when using `Trigger`).

I also fixed a CSS issue with the cursor being diaplayed as `pointer` on the whole screen.

TomasEng commented:
I've now pushed a new fix, setting the file changes info modal width to `75vw`. Since the height is `75vh`, it now takes up the same fraction of the width as of the height.

standeren reviewed (approved):
Tested OK in both safari and chrome 💪
